cmds-42/usr.etc/nmserver/man/nmxlog.1
.TH NMXLOG 1 2/22/88
.CM 4
.SH NAME
nmxlog \- translate a Mach network server debugging log
.SH SYNOPSIS
.B nmxlog
[ -d dictionary ] [ infile ]
.SH DESCRIPTION
.I nmxlog
translates the debugging log from a Mach network server from the internal
binary format into a human-readable format.

The dictionary file consists of a series of lines, beginning with the
event code in the log, followed by a 
format string suitable for use with
.I sprintf
.li
.
Each format specification is replaced with a value from the log record 
for the event.
The dictionary file is searched first in the location specified
with the 
.B -d
switch if present, then in the location specified by the 
.B NMLOGDICT
environment variable, and finally in the standard location
.I /usr/mach/lib/net_dict.data
.li
.

If no input file is specified,
.I stdin
is used.
The output always goes to
.I stdout.
.SH FILES
.TP 
/usr/mach/lib/net_dict.data       standard dictionary file.
.i0
.DT
.PP
.SH ENVIRONMENT
.TP 
NMLOGDICT                         location of dictionary file.
.i0
.DT
.PP
.SH "SEE ALSO"
nmcontrol(1), nmserver(8)
.br
"Network Server Design", internal document from the Mach group.
.SH HISTORY
.TP 
22-Feb-88  Daniel Julin (dpj) at Carnegie-Mellon University
Created.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.etc/nmserver/utils/nmxlog.c
/* 
 * Mach Operating System
 * Copyright (c) 1987 Carnegie-Mellon University
 * All rights reserved.  The CMU software License Agreement specifies
 * the terms and conditions for use and redistribution.
 */
/*
 * nmxlog.c
 *
 * $Source: /os/osdev/LIBS/libs-7/etc/nmserver/utils/RCS/nmxlog.c,v $
 *
 */
#ifndef	lint
static char     rcsid[] = "$Header: nmxlog.c,v 1.1 88/09/30 15:47:14 osdev Exp $";
#endif not lint
/*
 * Program to translate a log from a network server.
 * Input = stdin or file argument.
 * Output = stdout.
 * Dictionary obtain (in order) from: 
 *	-d switch on the command line
 *	environment variable "NMLOGDICT"
 *	standard file specified by STDDICTFILE
 */

/*
 * HISTORY: 
 * 13-Apr-87  Robert Sansom (rds) at Carnegie Mellon University
 *	Replaced net_recursion_level by trace_recursion_level
 *	and net_trace tracing_on.  Removed net_debug_level.
 *
 * 21-Mar-87  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Created.
 *
 */


#include	"trace.h"
#include	"nm_defs.h"
#include	"ls_defs.h"
#include	<stdio.h>
#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/file.h>


/*
 * Tracing values.
 */
int	trace_recursion_level = 0;
#undef	tracing_on
int	tracing_on = 0;


/*
 * Default path to the dictionary file.
 */
#define	STDDICTFILE	"/usr/mach/lib/net_dict.data"

/*
 * Number of possible dictionary entries.
 */
#define	DICTMAX		10000

/*
 * Usage string.
 */
#define	USAGE		"Usage: %s [-d dictfile] [<] infile > outfile\n"

/*
 * Table of translations.
 */
typedef struct {
    	int	string;		/* is this a string-type entry ? */
	char	*text;		/* format for the entry */
} translation_t;
translation_t	 translations[DICTMAX];
char	*textptr;		/* current free location to use for text */

/*
 * read_dict
 *
 * Read the dictionary file into the translations table.
 *
 * Parameters:
 *
 * dict_file: pathname of dictionary file.
 *
 * Results: none.
 *
 * Side effects:
 *
 * Opens the dictionary file. Fills the translations table. May abort the program
 * in case of error.
 *
 * Design:
 *
 * Note:
 *
 */
read_dict(dict_file)
char	*dict_file;
BEGIN("read_dict")
	int	i;
	FILE	*dict_str;	/* stream for dict_file */
	struct stat	stat;
	int	code;
	int	last_code = -1;

	/*
	 * First zero the translations.
	 */
	for (i = 0; i < DICTMAX; i++)
		translations[i].text = NULL;

	/*
	 * Try to open the dictionary file.
	 */
	if (!dict_file) {
	    	if ((dict_file = (char *)getenv("NMLOGDICT")) == NULL) {
		    	dict_file = STDDICTFILE;
		}
	}
    	if ((dict_str = fopen(dict_file, "r")) == NULL) {
	    	fprintf(stderr,"Cannot open \"%s\"\n", dict_file);
		exit(1);
	}

	/*
	 * Allocate space for the translation text.
	 */
	if (fstat(fileno(dict_str),&stat)) {
	    	perror("Cannot stat dictionary file");
		exit(1);
	}
	textptr = (char *)malloc(stat.st_size);
	if (textptr == 0) {
	    	fprintf(stderr,"Cannot allocate space for the translation text\n");
		exit(1);
	}

	/*
	 * Load all the translations.
	 */
	while(!feof(dict_str)) {
	    	if (fscanf(dict_str,"%d",&code) != 1) {
		    	if (feof(dict_str))
				break;
		    	fprintf(stderr,"Invalid entry in dictionary file - last valid code = %d\n",last_code);
			exit(1);
		}
		if ((code < 0) || (code >= DICTMAX)) {
		    	fprintf(stderr,"Invalid code in dictionary file: %d\n",code);
			exit(1);
		}
		if (translations[code].text) {
		    	fprintf(stderr,"Duplicate entry in dictionary for code %d\n",code);
			exit(1);
		}
		if ((translations[code].text = fgets(textptr,200,dict_str)) == NULL) {
		    	fprintf(stderr,"Invalid text in dictionary file - code = %\n",code);
			exit(1);
		}
		translations[code].string = sindex(textptr,"%s");
		last_code = code;
		textptr += strlen(textptr) + 1;
	}

	/*
	 * Close the dictionary file.
	 */
	fclose(dict_str);

	RET;
END



/*
 * main
 *
 * Main loop.
 *
 * Parameters:
 *
 * Results:
 *
 * Side effects:
 *
 * Design:
 *
 * Note:
 *
 */
void main(argc,argv)
int	argc;
char	*argv[];
BEGIN("main")
	char	*in_file = NULL;		/* input file */
	char	*dict_file = NULL;		/* path to dictionary file */
	char	*prog;				/* name under which invoked */
	FILE	*istr = NULL;			/* input stream */
	FILE	*ostr;				/* output stream */
	log_rec_t	log_rec[4];		/* space to read log records into */
	int	code;				/* current log code */
	translation_t	*tr_ptr;		/* current translation record */
	char	strtmp[1000];			/* temporary string for translation */

	/*
	 * Parse the command line.
	 */
	prog = *argv;
	for (argv++; --argc; argv++) {
	    	if ((*argv)[0] == '-') {
		    	switch((*argv)[1]) {
			    	case 'd':
					if (argc < 2) {
					    	fprintf(stderr, "No file name after \"-d\"\n");
					    	fprintf(stderr, USAGE, prog);
						exit(1);
					}
					argv++;
					argc--;
					dict_file = *argv;
					break;
				default:
					fprintf(stderr, "Unknown switch\n");
				    	fprintf(stderr, USAGE, prog);
					exit(1);
					break;
			}
		} else {
		    	if (in_file) {
			    	fprintf(stderr, "Multiple input specification\n");
			    	fprintf(stderr, USAGE, prog);
				exit(1);
			}
		    	in_file = *argv;
		}
	}

	/*
	 * Read the dictionary file.
	 */
	 read_dict(dict_file);

	/*
	 * Worry about input and output.
	 */
	if (in_file) {
	    	if ((istr = fopen(in_file, "r")) == NULL) {
		    	fprintf(stderr,"Cannot open \"%s\"\n", in_file);
			exit(1);
		}
	} else {
	    	istr = stdin;
	}
	ostr = stdout;

	/*
	 * Read and translate all log records.
	 */
	while(fread(log_rec,sizeof(log_rec_t),1,istr)) {
	    	code = log_rec[0].code;
		if ((code < 0) || (code >= DICTMAX) ||
			((tr_ptr = &translations[log_rec[0].code])->text == NULL)) {
		    	fprintf(stderr,"Invalid code: %d\n",log_rec[0].code);
		    	fprintf(ostr,"(%d) %d : --- Invalid code %x %x %x %x %x %x\n",
				log_rec[0].thread,
				log_rec[0].code,
				log_rec[0].a1,
				log_rec[0].a2,
				log_rec[0].a3,
				log_rec[0].a4,
				log_rec[0].a5,
				log_rec[0].a6);
			continue;
		}
	    	if (tr_ptr->string) {
		    	if (fread(&log_rec[1],sizeof(log_rec_t),3,istr) == 0) {
			    	fprintf(stderr,"Premature EOF on log string entry\n");
				break;
			}
			sprintf(strtmp,tr_ptr->text,&log_rec[0].a1);
		} else {
		    	sprintf(strtmp,tr_ptr->text,
				log_rec[0].a1,
				log_rec[0].a2,
				log_rec[0].a3,
				log_rec[0].a4,
				log_rec[0].a5,
				log_rec[0].a6);
		}
		fprintf(ostr,"(%6d) %4d : %s",log_rec[0].thread,log_rec[0].code,strtmp);
	}

	RET;
END
